File: backend/app/api/routes/security_events.py
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.api.deps import (
    get_active_membership,
    get_db,
    require_permission,
)
from app.db.models.membership import Membership
from app.schemas.security_event import (
    SecurityEventCreate,
    SecurityEventList,
    SecurityEventRead,
    SecurityEventUpdate,
)
from app.services.security_event_service import SecurityEventService


router = APIRouter(
    prefix="/security-events",
    tags=["Security Events"],
)


def get_security_service(
    db: Session = Depends(get_db),
) -> SecurityEventService:
    return SecurityEventService(db)


def event_not_found() -> HTTPException:
    return HTTPException(
        status_code=status.HTTP_404_NOT_FOUND,
        detail="Security event not found.",
    )


# ----------------------------------------------------------------------
# Create
# ----------------------------------------------------------------------

@router.post(
    "",
    response_model=SecurityEventRead,
    status_code=status.HTTP_201_CREATED,
    dependencies=[
        Depends(require_permission("security.events.create")),
    ],
)
def create_security_event(
    payload: SecurityEventCreate,
    membership: Membership = Depends(get_active_membership),
    service: SecurityEventService = Depends(get_security_service),
) -> SecurityEventRead:
    return service.create(
        tenant_id=membership.tenant_id,
        payload=payload,
    )


# ----------------------------------------------------------------------
# List
# ----------------------------------------------------------------------

@router.get(
    "",
    response_model=SecurityEventList,
    dependencies=[
        Depends(require_permission("security.events.view")),
    ],
)
def list_security_events(
    q: str | None = Query(
        default=None,
        max_length=100,
        description="Search across normalized event fields.",
    ),
    severity: str | None = Query(
        default=None,
        pattern="^(info|low|medium|high|critical)$",
    ),
    event_status: str | None = Query(
        default=None,
        alias="status",
        pattern="^(open|processing|processed|failed|suppressed)$",
    ),
    event_category: str | None = Query(
        default=None,
        max_length=100,
    ),
    event_type: str | None = Query(
        default=None,
        max_length=150,
    ),
    source: str | None = Query(
        default=None,
        max_length=255,
    ),
    source_type: str | None = Query(
        default=None,
        max_length=100,
    ),
    hostname: str | None = Query(
        default=None,
        max_length=255,
    ),
    user_identifier: str | None = Query(
        default=None,
        max_length=255,
    ),
    mitre_technique_id: str | None = Query(
        default=None,
        max_length=50,
    ),
    start_time: datetime | None = Query(
        default=None,
        description="Return events occurring on or after this timestamp.",
    ),
    end_time: datetime | None = Query(
        default=None,
        description="Return events occurring on or before this timestamp.",
    ),
    limit: int = Query(
        default=50,
        ge=1,
        le=500,
    ),
    offset: int = Query(
        default=0,
        ge=0,
    ),
    membership: Membership = Depends(get_active_membership),
    service: SecurityEventService = Depends(get_security_service),
) -> SecurityEventList:
    if (
        start_time is not None
        and end_time is not None
        and start_time > end_time
    ):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="start_time must be earlier than or equal to end_time.",
        )

    items, total = service.list(
        tenant_id=membership.tenant_id,
        limit=limit,
        offset=offset,
        q=q,
        severity=severity,
        status=event_status,
        event_category=event_category,
        event_type=event_type,
        source=source,
        source_type=source_type,
        hostname=hostname,
        user_identifier=user_identifier,
        mitre_technique_id=mitre_technique_id,
        start_time=start_time,
        end_time=end_time,
    )

    return SecurityEventList(
        items=items,
        total=total,
        limit=limit,
        offset=offset,
    )


# ----------------------------------------------------------------------
# Get
# ----------------------------------------------------------------------

@router.get(
    "/{event_id}",
    response_model=SecurityEventRead,
    dependencies=[
        Depends(require_permission("security.events.view")),
    ],
)
def get_security_event(
    event_id: UUID,
    membership: Membership = Depends(get_active_membership),
    service: SecurityEventService = Depends(get_security_service),
) -> SecurityEventRead:
    event = service.get(
        tenant_id=membership.tenant_id,
        event_id=event_id,
    )

    if event is None:
        raise event_not_found()

    return event


# ----------------------------------------------------------------------
# Update
# ----------------------------------------------------------------------

@router.patch(
    "/{event_id}",
    response_model=SecurityEventRead,
    dependencies=[
        Depends(require_permission("security.events.update")),
    ],
)
def update_security_event(
    event_id: UUID,
    payload: SecurityEventUpdate,
    membership: Membership = Depends(get_active_membership),
    service: SecurityEventService = Depends(get_security_service),
) -> SecurityEventRead:
    if not payload.model_dump(exclude_unset=True):
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail="At least one field is required.",
        )

    event = service.update(
        tenant_id=membership.tenant_id,
        event_id=event_id,
        payload=payload,
    )

    if event is None:
        raise event_not_found()

    return event


# ----------------------------------------------------------------------
# Delete
# ----------------------------------------------------------------------

@router.delete(
    "/{event_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[
        Depends(require_permission("security.events.delete")),
    ],
)
def delete_security_event(
    event_id: UUID,
    membership: Membership = Depends(get_active_membership),
    service: SecurityEventService = Depends(get_security_service),
) -> None:
    deleted = service.delete(
        tenant_id=membership.tenant_id,
        event_id=event_id,
    )

    if not deleted:
        raise event_not_found()

    return None

File: frontend/lib/api/security-events.ts
import { authenticatedFetch } from './client'

const SECURITY_EVENTS_ENDPOINT = '/api/security-events'

function tenantSecurityEventsEndpoint(tenantId?: string): string {
  return tenantId
    ? `/api/v1/tenants/${encodeURIComponent(tenantId)}/security-events`
    : SECURITY_EVENTS_ENDPOINT
}

export type SecurityEventSeverity =
  | 'info'
  | 'low'
  | 'medium'
  | 'high'
  | 'critical'

export type SecurityEventStatus =
  | 'open'
  | 'processing'
  | 'processed'
  | 'failed'
  | 'suppressed'

type JsonObject = Record<string, unknown>

export interface SecurityEvent {
  readonly id: string
  readonly tenant_id: string

  readonly source_event_id?: string | null
  readonly event_fingerprint?: string | null
  readonly correlation_id?: string | null
  readonly parent_event_id?: string | null

  readonly event_time: string
  readonly ingested_at: string
  readonly created_at: string
  readonly schema_version: number

  readonly event_category?: string | null
  readonly source: string
  readonly source_type: string
  readonly event_type: string

  readonly severity: SecurityEventSeverity
  readonly status: SecurityEventStatus

  readonly action?: string | null
  readonly risk_score?: number | null

  readonly source_ip?: string | null
  readonly destination_ip?: string | null
  readonly source_port?: number | null
  readonly destination_port?: number | null
  readonly protocol?: string | null

  readonly hostname?: string | null
  readonly user_identifier?: string | null
  readonly process_name?: string | null
  readonly process_id?: number | null

  readonly mitre_tactic?: string | null
  readonly mitre_technique?: string | null
  readonly mitre_technique_id?: string | null

  readonly message?: string | null

  readonly raw_event?: JsonObject | null
  readonly normalized_data?: JsonObject | null
  readonly event_metadata?: JsonObject | null
}

export interface SecurityEventCreate {
  readonly source_event_id?: string | null
  readonly event_fingerprint?: string | null
  readonly correlation_id?: string | null
  readonly parent_event_id?: string | null

  readonly event_time: string
  readonly schema_version?: number

  readonly event_category?: string | null
  readonly source: string
  readonly source_type: string
  readonly event_type: string

  readonly severity?: SecurityEventSeverity
  readonly status?: SecurityEventStatus

  readonly action?: string | null
  readonly risk_score?: number | null

  readonly source_ip?: string | null
  readonly destination_ip?: string | null
  readonly source_port?: number | null
  readonly destination_port?: number | null
  readonly protocol?: string | null

  readonly hostname?: string | null
  readonly user_identifier?: string | null
  readonly process_name?: string | null
  readonly process_id?: number | null

  readonly mitre_tactic?: string | null
  readonly mitre_technique?: string | null
  readonly mitre_technique_id?: string | null

  readonly message?: string | null

  readonly raw_event?: JsonObject | null
  readonly normalized_data?: JsonObject | null
  readonly event_metadata?: JsonObject | null
}

export interface SecurityEventListResponse {
  readonly items: SecurityEvent[]
  readonly total: number
  readonly limit: number
  readonly offset: number
}

export interface SecurityEventFilters {
  readonly q?: string
  readonly severity?: SecurityEventSeverity
  readonly status?: SecurityEventStatus
  readonly event_category?: string
  readonly event_type?: string
  readonly source?: string
  readonly source_type?: string
  readonly hostname?: string
  readonly user_identifier?: string
  readonly mitre_technique_id?: string
  readonly start_time?: string
  readonly end_time?: string
  readonly limit?: number
  readonly offset?: number
}

export interface SecurityEventUpdate {
  readonly status?: SecurityEventStatus
  readonly event_metadata?: JsonObject | null
}

export class ApiError extends Error {
  constructor(
    public readonly status: number,
    message: string,
  ) {
    super(message)
    this.name = 'ApiError'
  }
}

function isRecord(value: unknown): value is JsonObject {
  return typeof value === 'object' && value !== null
}

function getEventPath(eventId: string): string {
  return `/${encodeURIComponent(eventId)}`
}

function buildQueryString(filters: SecurityEventFilters = {}): string {
  const params = new URLSearchParams()

  const queryFilters: Record<string, unknown> = {
    q: filters.q,
    severity: filters.severity,
    status: filters.status,
    event_category: filters.event_category,
    event_type: filters.event_type,
    source: filters.source,
    source_type: filters.source_type,
    hostname: filters.hostname,
    user_identifier: filters.user_identifier,
    mitre_technique_id: filters.mitre_technique_id,
    start_time: filters.start_time,
    end_time: filters.end_time,
    limit: filters.limit,
    offset: filters.offset,
  }

  for (const [key, value] of Object.entries(queryFilters)) {
    if (value === undefined || value === null) {
      continue
    }

    if (typeof value === 'string') {
      const normalizedValue = value.trim()

      if (!normalizedValue) {
        continue
      }

      params.set(key, normalizedValue)
      continue
    }

    params.set(key, String(value))
  }

  const query = params.toString()

  return query ? `?${query}` : ''
}

async function parseErrorMessage(response: Response): Promise<string> {
  const fallback = `API_ERROR_${response.status}`
  const contentType = response.headers.get('content-type') ?? ''

  if (!contentType.includes('application/json')) {
    return fallback
  }

  try {
    const body: unknown = await response.json()

    if (!isRecord(body) || body.detail === undefined) {
      return fallback
    }

    if (typeof body.detail === 'string') {
      return body.detail
    }

    return JSON.stringify(body.detail) || fallback
  } catch {
    return fallback
  }
}

async function handleResponse<T>(response: Response): Promise<T> {
  if (response.status === 204) {
    return undefined as T
  }

  const contentType = response.headers.get('content-type') ?? ''
  const isJson = contentType.includes('application/json')

  if (!response.ok) {
    throw new ApiError(response.status, await parseErrorMessage(response))
  }

  if (!isJson) {
    throw new ApiError(
      response.status,
      'The API returned an unexpected response format.',
    )
  }

  try {
    return (await response.json()) as T
  } catch {
    throw new ApiError(response.status, 'The API returned invalid JSON.')
  }
}

async function apiFetch(
  tenantId: string | undefined,
  endpoint = '',
  options: RequestInit = {},
): Promise<Response> {
  const headers = new Headers(options.headers)

  headers.set('Accept', 'application/json')

  if (options.body !== undefined && !headers.has('Content-Type')) {
    headers.set('Content-Type', 'application/json')
  }

  return authenticatedFetch(`${tenantSecurityEventsEndpoint(tenantId)}${endpoint}`, {
    ...options,
    headers,
  })
}

function hasUpdateFields(updates: SecurityEventUpdate): boolean {
  return Object.values(updates).some((value) => value !== undefined)
}

export async function getSecurityEvents(
  filters: SecurityEventFilters = {},
  tenantId?: string,
): Promise<SecurityEventListResponse> {
  const response = await apiFetch(tenantId, buildQueryString(filters))

  return handleResponse<SecurityEventListResponse>(response)
}

export async function getSecurityEvent(
  eventId: string,
  tenantId?: string,
): Promise<SecurityEvent> {
  const response = await apiFetch(tenantId, getEventPath(eventId))

  return handleResponse<SecurityEvent>(response)
}

export async function createSecurityEvent(
  event: SecurityEventCreate,
  tenantId?: string,
): Promise<SecurityEvent> {
  const response = await apiFetch(tenantId, '', {
    method: 'POST',
    body: JSON.stringify(event),
  })

  return handleResponse<SecurityEvent>(response)
}

export async function updateSecurityEvent(
  eventId: string,
  updates: SecurityEventUpdate,
  tenantId?: string,
): Promise<SecurityEvent> {
  if (!hasUpdateFields(updates)) {
    throw new TypeError('At least one defined update field is required.')
  }

  const response = await apiFetch(tenantId, getEventPath(eventId), {
    method: 'PATCH',
    body: JSON.stringify(updates),
  })

  return handleResponse<SecurityEvent>(response)
}

export async function deleteSecurityEvent(eventId: string, tenantId?: string): Promise<void> {
  const response = await apiFetch(tenantId, getEventPath(eventId), {
    method: 'DELETE',
  })

  await handleResponse<void>(response)
}
